src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -1296,25 +1210,61 @@ void ObjectMonitor::exit(JavaThread* current, bool not_suspended) {
-    // shorter than the worst-case round-trip cache-line migration time.
-    // The dropped lock needs to become visible to the spinner, and then
-    // the acquisition of the lock by the spinner must become visible to
-    // the exiting thread).
+    // Which means that the exiting thread could exit immediately without
+    // waking a successor, if it observes a successor after it has dropped
+    // the lock.  Note that the dropped lock needs to become visible to the
+    // spinner.
 
     // It appears that an heir-presumptive (successor) must be made ready.
     // Only the current lock owner can manipulate the EntryList or
     // drain _cxq, so we need to reacquire the lock.  If we fail
     // to reacquire the lock the responsibility for ensuring succession
     // falls to the new owner.
     //
-    if (try_set_owner_from(nullptr, current) != nullptr) {
-      return;
+    void* owner = try_set_owner_from(nullptr, current);
+    if (owner != nullptr) {
+      if (owner != DEFLATER_MARKER) {
+        // Owned by another thread, so we are done.
+        return;
+      }
+
+      // The deflator owns the lock.  Now try to cancel the async
+      // deflation.  As part of the contended enter protocol,
+      // contentions was incremented to a positive value before
+      // EnterI() was called and that prevents the deflater thread
+      // from winning the last part of the 2-part async deflation
+      // protocol. After EnterI() returns to enter(), contentions is
+      // decremented because the caller now owns the monitor. We need
+      // to increment contentions an extra time here (done by the
+      // ObjectMonitorContentionMark constructor) to prevent the
+      // deflater thread from winning the last part of the 2-part
+      // async deflation protocol after the regular decrement occurs
+      // in enter().
+      ObjectMonitorContentionMark contention_mark(this);
+
+      if (is_being_async_deflated()) {
+        assert((intptr_t(_EntryList) | intptr_t(_cxq)) == 0 || _succ != nullptr, "");
+        // Mr. Deflator won the race, so we are done.
+        return;
+      }
+
+      // Now try to take the ownership.
+      if (try_set_owner_from(DEFLATER_MARKER, current) == DEFLATER_MARKER) {
+        // We successfully cancelled the in-progress async deflation.
+        // By extending the lifetime of the contention_mark, we
+        // prevent the destructor from decrementing the contentions
+        // counter when the contention_mark goes out of scope. Instead
+        // ObjectMonitor::deflate_monitor() will decrement contentions
+        // after it recognizes that the async deflation was cancelled.
+        contention_mark.extend();
+      } else {
+        owner = try_set_owner_from(nullptr, current);
+        if (owner != nullptr) {
+          // The lock is owned by another thread, who is now
+          // responsible for ensuring succession, so we are done.
+          return;
+        }
+      }
```

Comment:
Can't we replace all this code for a call to TryLock?

Reply:
I think we can. Thanks for the suggestion.

Reply:
fixed